### PR TITLE
Script library

### PR DESCRIPTION
**Pull request by litlfred**
This is a pull request for @mortenoh @larshelge to perform a code review against.

This provides support for "Server Side Apps" and support for the FHIR standard.

Example "hello world" server side apps  are found here:
  https://github.com/openhie/dhis-ssa-hello-world
Once installed, it can be accessed at:
  http://localhost:8080/api/apps/dhis-ssa-hello-world/ssa/index.js

The beginnings of a convenience wrapper for core DHIS2 services is found here:
  https://github.com/openhie/dhis-ssa-core-services

FHIR support apps are being actively worked on here:
  https://github.com/openhie/dhis-ssa-metadata-fhir
  https://github.com/openhie/dhis-ssa-fhir-terminologies

### CODE REVIEW

**Comment by litlfred** [comment]
Hi,
Any update on the code review?
Cheers,
-carl

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/appmanager/AppManager.java [hunk 1]
```diff
@@ -50,22 +55,22 @@
...
+    List<App> getAppsByType ( AppType appType, Collection<App> apps );
 
     /**
      * Returns a list of all installed apps with name equal the given name
      * and operator. Currently supports eq and ilike.
      *
      * @return list of installed apps with given name
      */
-    List<App> getAppsByName( String name, Collection<App> apps, String operator );
+    List<App> getAppsByName ( String name, Collection<App> apps, String operator );
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/appmanager/AppManager.java]: Follow DHIS2 code style. No space after method name.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/scriptlibrary/EngineInterface.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+import java.io.Reader;
+import java.util.concurrent.Callable;
+
+
+public interface  EngineInterface extends Callable
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/scriptlibrary/EngineInterface.java]: Don't name an interface "interface", redundant. Give more descriptive name.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/scriptlibrary/EngineInterface.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+{
+
+    abstract Object evaluateScript() throws ScriptException;
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/scriptlibrary/EngineInterface.java]: Pointless to use abstract modifier in interfaces, remove.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/scriptlibrary/ExecutionContext.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    public void setError ( Writer error )
+    {
+        this.error = error;
+    }
+
+    public void logInfo(Object o) {
+        log.info(o);
+    }
+
+    public void logDebug(Object o) {
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/scriptlibrary/ExecutionContext.java]: DHIS2 code style needed.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/scriptlibrary/ExecutionContextInterface.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+public interface ExecutionContextInterface {
+
+    abstract String getAppKey();
+
+    abstract String getScriptName();
+
+    abstract User getUser();
+
+    abstract Reader getIn();
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/scriptlibrary/ExecutionContextInterface.java]: Remove abstract modifier.

**Comment by litlfred** [comment]
Hi,
Can we talk about a review process to take this forward?
Cheers,
-carl

**Comment by litlfred** [comment]
Adding in a link to the documentation that was generated when specifying the requirements for this:

https://docs.google.com/document/d/1eEnRI8qx6HRMYw3On_gEsMI5v6K_uMj7LoZo2GnSRsA/edit#heading=h.dagobvz96naz

@larshelge @mortenoh @olatitle

**Comment by larshelge** [comment]
Closing this PR for now.

**Pull request closed without merging**